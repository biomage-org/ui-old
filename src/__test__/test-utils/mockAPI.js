import _ from 'lodash';

// Default platform data
import cellSetsData from '__test__/data/cell_sets.json';
import backendStatusData from '__test__/data/backend_status.json';
import processingConfigData from '__test__/data/processing_config.json';
import mockDemoExperiments from '__test__/test-utils/mockData/mockDemoExperiments.json';

// A ticket has been created to address this : https://biomage.atlassian.net/browse/BIOMAGE-1553
import {
  responseData,
} from '__test__/test-utils/mockData';

const promiseResponse = (
  response,
  options = {},
) => Promise.resolve(new Response(response, options));

const statusResponse = (code, body) => Promise.resolve({
  status: code,
  body: JSON.stringify(body),
});

const delayedResponse = (response, delay = 10000, options = {}) => new Promise((resolve) => {
  setTimeout(() => {
    resolve(new Response(response, options));
  }, delay);
});

const workerResponse = (body) => promiseResponse(body);

const generateDefaultMockAPIResponses = (experimentId) => ({
  [`experiments/${experimentId}`]: () => promiseResponse(
    JSON.stringify(responseData.experiments.find(({ id }) => id === experimentId)),
  ),
  [`experiments/${experimentId}/processingConfig`]: () => promiseResponse(
    JSON.stringify(processingConfigData),
  ),
  [`experiments/${experimentId}/cellSets`]: () => promiseResponse(
    JSON.stringify(cellSetsData),
  ),
  [`experiments/${experimentId}/backendStatus`]: () => promiseResponse(
    JSON.stringify(backendStatusData),
  ),
  experiments: () => promiseResponse(
    JSON.stringify(responseData.experiments),
  ),
  [`experiments/${experimentId}/samples`]: () => promiseResponse(
    JSON.stringify(responseData.samples[0]),
  ),
  '/v2/experiments/examples': () => promiseResponse(
    JSON.stringify(mockDemoExperiments),
  ),
});

const mockAPI = (apiMapping) => (req) => {
  const path = req.url;

  const key = _.find(
    Object.keys(apiMapping),
    (urlStub) => path.endsWith(urlStub),
  );

  if (!key) {
    return statusResponse({
      status: 404,
      body: `Path ${path} is undefined`,
    });
  }

  return apiMapping[key](req);
};

export default mockAPI;
export {
  generateDefaultMockAPIResponses,
  promiseResponse,
  statusResponse,
  workerResponse,
  delayedResponse,
};
