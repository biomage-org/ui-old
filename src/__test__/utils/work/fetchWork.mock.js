const mockCacheKeyMappings = {
  A: 'fd3161a878f67ebf54018720cffd6a66', // pragma: allowlist secret
  B: '10250a11679234110a1c260d6fd81d3c', // pragma: allowlist secret
  C: '76bf160685c4c80c67abd9a701da23e6', // pragma: allowlist secret
  D: '21671038b9ac73c1f08a94c8213cb872', // pragma: allowlist secret
  E: '33faa711a94a2028b5bae1778126aec0', // pragma: allowlist secret
};

const mockGenesListData = ['A', 'B', 'C', 'D'];

const mockGeneExpressionData = {
  A: {
    rawExpression: {
      min: 0,
      max: 6.8,
      mean: 1.68,
      stdev: 2.597331964,
      expression: [0, 0, 0, 1.56, 0, 6.8, 3.4],
    },
    truncatedExpression: {
      min: 0,
      max: 6.8,
      mean: 1.68,
      stdev: 2.597331964,
      expression: [0, 0, 0, 1.56, 0, 6.8, 3.4],
    },
  },
  B: {
    rawExpression: {
      min: 0,
      max: 6.8,
      mean: 1.702857143,
      stdev: 2.551115536,
      expression: [0, 0, 0, 2.56, 0, 6.8, 2.56],
    },
    truncatedExpression: {
      min: 0,
      max: 6.8,
      mean: 1.702857143,
      stdev: 2.551115536,
      expression: [0, 0, 0, 2.56, 0, 6.8, 2.56],
    },
  },
  C: {
    rawExpression: {
      min: 0,
      max: 3.4,
      mean: 1.68,
      stdev: 2.141525936,
      expression: [0, 0, 0, 3.56, 0, 4.8, 3.4],
    },
    truncatedExpression: {
      min: 0,
      max: 3.4,
      mean: 1.68,
      stdev: 2.141525936,
      expression: [0, 0, 0, 3.56, 0, 4.8, 3.4],
    },
  },
  D: {
    rawExpression: {
      min: 0,
      max: 6.8,
      mean: 1.68,
      stdev: 2.597331964,
      expression: [0, 0, 0, 1.56, 0, 6.8, 3.4],
    },
    truncatedExpression: {
      min: 0,
      max: 6.8,
      mean: 1.68,
      stdev: 2.597331964,
      expression: [0, 0, 0, 1.56, 0, 6.8, 3.4],
    },
  },
  E: { hello: 'world' },
};

const mockGeneExpressionDataZScore = {
  D: {
    zScore: [
      -0.6468175894669735,
      -0.6468175894669735,
      -0.6468175894669735,
      -0.04620125639049807,
      -0.6468175894669735,
      1.971253605994586,
      0.6622180082638063,
    ],
  },
};

const mockCacheGet = jest.fn((x) => {
  const mockCacheContents = {
    fd3161a878f67ebf54018720cffd6a66: 'A', // pragma: allowlist secret
    '10250a11679234110a1c260d6fd81d3c': 'B', // pragma: allowlist secret
    f5c957411a28de68f35e1f5c8a29da7e: 'C', // pragma: allowlist secret
    '33faa711a94a2028b5bae1778126aec0': 'E', // pragma: allowlist secret
  };

  if (x in mockCacheContents) {
    return mockGeneExpressionData[mockCacheContents[x]];
  }

  return null;
});
const mockCacheSet = jest.fn();
const mockCacheRemove = jest.fn();

const mockDispatchWorkRequest = jest.fn(() => true);

const mockSeekFromS3 = jest.fn();

const mockCacheModule = {
  get: jest.fn((x) => mockCacheGet(x)),
  set: jest.fn((key, val) => mockCacheSet(key, val)),
  _remove: jest.fn((key) => mockCacheRemove(key)),
};

const mockSeekWorkResponseModule = {
  __esModule: true,
  seekFromS3: mockSeekFromS3,
  dispatchWorkRequest: mockDispatchWorkRequest,
};

const mockQcPipelineStartDate = '2021-01-01T01:01:01.000Z';

const mockReduxState = (experimentId, environment = 'testing') => () => ({
  networkResources: {
    environment,
  },
  backendStatus: {
    [experimentId]: {
      status: {
        pipeline: {
          status: 'SUCCEEDED',
          startDate: mockQcPipelineStartDate,
        },
      },
    },
  },
});

export {
  mockGeneExpressionData, mockGeneExpressionDataZScore,
  mockGenesListData,
  mockCacheKeyMappings,
  mockCacheGet, mockCacheSet,
  mockDispatchWorkRequest, mockSeekFromS3,
  mockReduxState, mockQcPipelineStartDate,
  mockCacheModule, mockSeekWorkResponseModule,
};
