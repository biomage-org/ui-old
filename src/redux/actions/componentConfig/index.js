import loadPlotConfig from 'redux/actions/componentConfig/loadPlotConfig';
import loadComponentConfig from 'redux/actions/componentConfig/loadComponentConfig';
import updatePlotConfig from 'redux/actions/componentConfig/updatePlotConfig';
import savePlotConfig from 'redux/actions/componentConfig/savePlotConfig';
import getDotPlot from 'redux/actions/componentConfig/getDotPlot';
import updatePlotData from 'redux/actions/componentConfig/updatePlotData';
import resetPlotConfig from 'redux/actions/componentConfig/resetPlotConfig';

export {
  loadPlotConfig,
  loadComponentConfig,
  updatePlotConfig,
  savePlotConfig,
  getDotPlot,
  updatePlotData,
  resetPlotConfig,
};
