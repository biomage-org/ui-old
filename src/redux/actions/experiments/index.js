import createExperiment from './createExperiment';
import deleteExperiment from './deleteExperiment';
import updateExperiment from './updateExperiment';
import loadExperiments from './loadExperiments';
import setActiveExperiment from './setActiveExperiment';
import switchExperiment from './switchExperiment';
import reorderSamples from './reorderSamples';
import createMetadataTrack from './createMetadataTrack';
import renameMetadataTrack from './renameMetadataTrack';
import deleteMetadataTrack from './deleteMetadataTrack';
import updateValueInMetadataTrack from './updateValueInMetadataTrack';
import cloneExperiment from './cloneExperiment';
import reloadExperimentInfo from './reloadExperimentInfo';

export {
  createExperiment,
  deleteExperiment,
  updateExperiment,
  loadExperiments,
  setActiveExperiment,
  switchExperiment,
  reorderSamples,
  createMetadataTrack,
  renameMetadataTrack,
  deleteMetadataTrack,
  updateValueInMetadataTrack,
  cloneExperiment,
  reloadExperimentInfo,
};
