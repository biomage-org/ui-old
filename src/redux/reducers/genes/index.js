import initialState from './initialState';
import {
  GENES_PROPERTIES_LOADING, GENES_PROPERTIES_LOADED_PAGINATED, GENES_PROPERTIES_ERROR,
  GENES_SELECT, GENES_DESELECT,
  GENES_EXPRESSION_LOADING, GENES_EXPRESSION_LOADED, GENES_EXPRESSION_ERROR,
  MARKER_GENES_LOADING, MARKER_GENES_LOADED, MARKER_GENES_ERROR,
} from '../../actionTypes/genes';

import { EXPERIMENT_SETTINGS_QC_START } from '../../actionTypes/experimentSettings';

import genesPropertiesLoading from './genesPropertiesLoading';
import genesPropertiesLoadedPaginated from './genesPropertiesLoadedPaginated';
import genesPropertiesError from './genesPropertiesError';
import genesExpressionLoading from './genesExpressionLoading';
import genesExpressionLoaded from './genesExpressionLoaded';
import genesExpressionError from './genesExpressionError';

import markerGenesLoading from './markerGenesLoading';
import markerGenesError from './markerGenesError';
import markerGenesLoaded from './markerGenesLoaded';

import genesSelect from './genesSelect';
import genesDeselect from './genesDeselect';

const genesReducer = (state = initialState, action) => {
  switch (action.type) {
    case GENES_PROPERTIES_LOADING: {
      return genesPropertiesLoading(state, action);
    }
    case GENES_PROPERTIES_LOADED_PAGINATED: {
      return genesPropertiesLoadedPaginated(state, action);
    }
    case GENES_PROPERTIES_ERROR: {
      return genesPropertiesError(state, action);
    }
    case GENES_SELECT: {
      return genesSelect(state, action);
    }
    case GENES_DESELECT: {
      return genesDeselect(state, action);
    }
    case GENES_EXPRESSION_LOADING: {
      return genesExpressionLoading(state, action);
    }
    case GENES_EXPRESSION_LOADED: {
      return genesExpressionLoaded(state, action);
    }
    case GENES_EXPRESSION_ERROR: {
      return genesExpressionError(state, action);
    }
    case EXPERIMENT_SETTINGS_QC_START: {
      return initialState;
    }
    case MARKER_GENES_LOADING: {
      return markerGenesLoading(state, action);
    }
    case MARKER_GENES_LOADED: {
      return markerGenesLoaded(state, action);
    }
    case MARKER_GENES_ERROR: {
      return markerGenesError(state, action);
    }
    default: {
      return state;
    }
  }
};

export default genesReducer;
