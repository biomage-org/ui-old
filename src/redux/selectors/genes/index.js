import getGeneList from './getGeneList';

/* eslint-disable-next-line import/prefer-default-export */
export { getGeneList };
