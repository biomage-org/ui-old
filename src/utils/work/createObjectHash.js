import { MD5 } from 'object-hash';

const createObjectHash = (object) => MD5(object);

export default createObjectHash;
