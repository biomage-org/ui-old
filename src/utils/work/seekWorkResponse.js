import dayjs from 'dayjs';

import getAuthJWT from 'utils/getAuthJWT';
import fetchAPI from 'utils/http/fetchAPI';
import unpackResult from 'utils/work/unpackResult';
import parseResult from 'utils/work/parseResult';
import WorkTimeoutError from 'utils/errors/http/WorkTimeoutError';
import WorkResponseError from 'utils/errors/http/WorkResponseError';
import httpStatusCodes from 'utils/http/httpStatusCodes';

const throwResponseError = (response) => {
  throw new Error(`Error ${response.status}: ${response.text}`, { cause: response });
};

// getRemainingWorkerStartTime returns how many more seconds the worker is expected to
// need to be running with an extra 1 minute for a bit of leeway
const getRemainingWorkerStartTime = (creationTimestamp) => {
  const now = new Date();
  const creationTime = new Date(creationTimestamp);
  const elapsed = parseInt((now - creationTime) / (1000), 10); // gives second difference

  // we assume a worker takes up to 5 minutes to start
  const totalStartup = 5 * 60;
  const remainingTime = totalStartup - elapsed;
  // add an extra minute just in case
  return remainingTime + 60;
};

const seekFromS3 = async (ETag, experimentId, taskName) => {
  let response;
  try {
    const url = `/v2/workResults/${experimentId}/${ETag}`;

    response = await fetchAPI(url);
  } catch (e) {
    if (e.statusCode === httpStatusCodes.NOT_FOUND) {
      return null;
    }

    throw e;
  }

  const storageResp = await fetch(response.signedUrl);

  if (!storageResp.ok) {
    throwResponseError(storageResp);
  }

  const unpackedResult = await unpackResult(storageResp);
  const parsedResult = parseResult(unpackedResult, taskName);

  return parsedResult;
};

const dispatchWorkRequest = async (
  experimentId,
  body,
  timeout,
  ETag,
  requestProps,
) => {
  console.error('dispatching work request', body);
  const { default: connectionPromise } = await import('utils/socketConnection');
  const io = await connectionPromise;

  const timeoutDate = dayjs().add(timeout, 's').toISOString();
  const authJWT = await getAuthJWT();

  const request = {
    ETag,
    socketId: io.id,
    experimentId,
    ...(authJWT && { Authorization: `Bearer ${authJWT}` }),
    timeout: timeoutDate,
    body,
    ...requestProps,
  };

  const timeoutPromise = new Promise((resolve, reject) => {
    const id = setTimeout(() => {
      reject(new WorkTimeoutError(timeoutDate, request));
    }, timeout * 1000);

    io.on(`WorkerInfo-${experimentId}`, (res) => {
      const { response: { podInfo: { name, creationTimestamp, phase } } } = res;

      const extraTime = getRemainingWorkerStartTime(creationTimestamp);
      if (phase === 'Pending' && extraTime > 0) {
        console.log(`worker ${name} started at ${creationTimestamp}. Adding ${extraTime} seconds to timeout.`);
        clearTimeout(id);
        setTimeout(() => {
          reject(new WorkTimeoutError(timeoutDate, request));
        }, (timeout + extraTime) * 1000);
      }
    });
  });

  const responsePromise = new Promise((resolve, reject) => {
    io.on(`WorkResponse-${ETag}`, (res) => {
      const { response } = res;

      if (response.error) {
        const { errorCode, userMessage } = response;
        console.error(errorCode, userMessage);

        return reject(
          new WorkResponseError(errorCode, userMessage, request),
        );
      }

      // If no error, the response should be ready on S3.
      // In this case, return true
      return resolve();
    });
  });

  const result = Promise.race([timeoutPromise, responsePromise]);

  // TODO switch to using normal WorkRequest for v2 requests
  io.emit('WorkRequest-v2', request);

  return result;
};

export { dispatchWorkRequest, seekFromS3 };
