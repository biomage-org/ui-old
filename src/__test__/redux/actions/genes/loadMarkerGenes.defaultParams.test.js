import fetchMock, { enableFetchMocks } from 'jest-fetch-mock';
import mockAPI, {
  generateDefaultMockAPIResponses,
  promiseResponse,
} from '__test__/test-utils/mockAPI';

import { MD5 } from 'object-hash';
import _ from 'lodash';
import backendStatusData from '__test__/data/backend_status.json';
import createObjectHash from 'utils/work/createObjectHash';
import generateExperimentSettingsMock from '__test__/test-utils/experimentSettings.mock';
import { loadBackendStatus } from 'redux/actions/backendStatus';
import { loadMarkerGenes } from 'redux/actions/genes';
import loadProcessingSettings from 'redux/actions/experimentSettings/processingConfig/loadProcessingSettings';
import { makeStore } from 'redux/store';
import { seekFromS3 } from 'utils/work/seekWorkResponse';

jest.mock('utils/getTimeoutForWorkerTask', () => ({
  __esModule: true, // this property makes it work
  default: () => 60,
}));

enableFetchMocks();

jest.mock('utils/work/seekWorkResponse', () => ({
  __esModule: true, // this property makes it work
  dispatchWorkRequest: jest.fn(() => true),
  seekFromS3: jest.fn(),
}));

jest.mock('utils/work/createObjectHash');

const experimentId = '6463cb35-3e08-4e94-a181-6d155a5ca570';
const sampleIds = ['sample-WT', 'sample-WT1', 'sample-KO'];
const initialExperimentSettingsState = generateExperimentSettingsMock(sampleIds);

const experimentSettings = {
  ...initialExperimentSettingsState,
  processing: {
    ...initialExperimentSettingsState.processing,
    configureEmbedding: {
      clusteringSettings: {
        method: 'louvain',
        methodSettings: {
          louvain: {
            resolution: 0.8,
          },
        },
      },
      embeddingSettings: {
        method: 'umap',
        methodSettings: {
          tsne: { perplexity: 30, learningRate: 200 },
          umap: { distanceMetric: 'cosine', minimumDistance: 0.3 },
        },
      },
    },
  },
};

// this is the same Date used in the API to make sure the default ETag generated
// by the UI is identical to the API
const date = new Date(1458619200000);
backendStatusData.pipeline.startDate = date.toISOString();

const customAPIResponses = {
  [`experiments/${experimentId}/processingConfig`]: () => promiseResponse(JSON.stringify(experimentSettings.processing)),
  [`experiments/${experimentId}/backendStatus`]: () => promiseResponse(
    JSON.stringify(backendStatusData),
  ),
};

const mockApiResponses = _.merge(
  generateDefaultMockAPIResponses(experimentId), customAPIResponses,
);

describe('loadEmbedding action', () => {
  let store;

  beforeEach(async () => {
    jest.clearAllMocks();
    fetchMock.resetMocks();
    fetchMock.doMock();
    fetchMock.mockIf(/.*/, mockAPI(mockApiResponses));

    seekFromS3
      .mockReset()
      .mockImplementationOnce(() => null)
      .mockImplementationOnce(() => Promise.resolve([[1, 2], [3, 4]]));
    createObjectHash.mockImplementation((object) => MD5(object));

    store = makeStore();
    await store.dispatch(loadProcessingSettings(experimentId));
    await store.dispatch(loadBackendStatus(experimentId));
  });

  it('loadEmbedding generates correct hash params body for ETag', async () => {
    const hashMock = jest.fn((object) => MD5(object));
    createObjectHash.mockImplementation(hashMock);

    // eslint-disable-next-line max-len
    const { resolution } = experimentSettings.processing.configureEmbedding.clusteringSettings.methodSettings.louvain;

    await store.dispatch(loadMarkerGenes(
      experimentId,
      resolution,
    ));

    expect(hashMock).toHaveBeenCalled();
    // this snapshot should max exactly API snapshot:
    // submitMarkerHeatmapWork.test.js.snap
    expect(hashMock.mock.calls).toMatchSnapshot('marker heatmap - hash call params');
    // this ETag should match exactly the one in
    // submitMarkerHeatmap.test.js
    const ETag = hashMock.mock.results[0].value;
    expect(ETag).toEqual('9db473fff00ea358446196ee3276f486'); // pragma: allowlist secret`
  });
});
