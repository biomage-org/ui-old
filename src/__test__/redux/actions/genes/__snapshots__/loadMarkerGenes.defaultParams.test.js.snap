// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`loadEmbedding action loadEmbedding generates correct hash params body for ETag: marker heatmap - hash call params 1`] = `
[
  [
    {
      "body": {
        "cellSetKey": "louvain",
        "groupByClasses": [
          "louvain",
        ],
        "hiddenCellSetKeys": [],
        "nGenes": 5,
        "name": "MarkerHeatmap",
        "selectedPoints": "All",
      },
      "cacheUniquenessKey": null,
      "experimentId": "6463cb35-3e08-4e94-a181-6d155a5ca570",
      "extraDependencies": [
        {
          "method": "louvain",
          "methodSettings": {
            "louvain": {
              "resolution": 0.8,
            },
          },
        },
        [
          "louvain",
        ],
        1,
      ],
      "extras": undefined,
      "qcPipelineStartDate": "2016-03-22T04:00:00.000Z",
      "workerVersion": 3,
    },
  ],
]
`;
